### ISSUE TITLE
Empty "ENCRYPTION SETTINGS" panel with --log-format=CADDY and Caddy v2.7.4

### ISSUE THREAD

**Issue by steffenbusch** [allinurl/goaccess#2564]
I have a question/problem with GoAccess (latest Docker Image) and Caddy v2.7.4 regarding the "ENCRYPTION SETTINGS" panel, which does not show any data.

My setup is latest Caddy v2.7.4 and GoAccess using the Docker Image - following this useful guide https://twitter.com/ALEMRANCU/status/1642337601578471424 (retweeted on the goaccess twitter page)

`/var/log/caddy/goaccess.json --log-format=CADDY -o /var/www/goaccess/index.html --real-time-html --ws-url=wss://example.arm.stbu.net:443/ws --port=7890`

By default, the JSON access logs from Caddy as of v2.7.4 are like these:

```json
{"level":"info","ts":1696067106.8245707,"logger":"http.log.access.arm.stbu.net","msg":"handled request","request":{"remote_ip":"134.209.243.63","remote_port":"57526","client_ip":"134.209.243.63","proto":"HTTP/1.1","method":"GET","host":"arm.stbu.net","uri":"/","headers":{"Accept-Encoding":["gzip, deflate"],"User-Agent":["OhDear.app (+https://ohdear.app/docs/checks/uptime)"],"Accept":["*/*"]},"tls":{"resumed":false,"version":772,"cipher_suite":4865,"proto":"http/1.1","server_name":"arm.stbu.net"}},"bytes_read":0,"user_id":"","duration":0.00328789,"size":32,"status":302,"resp_headers":{"Server":["TheRock/2.0"],"Vary":["Accept"],"Content-Type":["text/plain; charset=utf-8"],"Content-Length":["32"],"Date":["Sat, 30 Sep 2023 09:45:06 GMT"],"X-Frame-Options":["SAMEORIGIN"],"Location":["/dashboard"]}}
```
So the `%K` should be `772` and `%k` should be `4865`.

Or with the human readable values output using this Caddyfile (using github.com/caddyserver/transform-encoder)

```txt
       log access-json {
                include http.log.access
                output file log/access.json
                format filter {
                        wrap json
                        fields {
                                request>tls>version tls_version TLSv
                                request>tls>cipher_suite tls_cipher
                        }
                }
        }
```

which produces JSON access logs like this:

```json
{"level":"info","ts":1696067106.8245707,"logger":"http.log.access.arm.stbu.net","msg":"handled request","request":{"remote_ip":"134.209.243.63","remote_port":"57526","client_ip":"134.209.243.63","proto":"HTTP/1.1","method":"GET","host":"arm.stbu.net","uri":"/","headers":{"Accept-Encoding":["gzip, deflate"],"User-Agent":["OhDear.app (+https://ohdear.app/docs/checks/uptime)"],"Accept":["*/*"]},"tls":{"resumed":false,"version":"TLSv1.3","cipher_suite":"TLS_AES_128_GCM_SHA256","proto":"http/1.1","server_name":"arm.stbu.net"}},"bytes_read":0,"user_id":"","duration":0.00328789,"size":32,"status":302,"resp_headers":{"Date":["Sat, 30 Sep 2023 09:45:06 GMT"],"X-Frame-Options":["SAMEORIGIN"],"Location":["/dashboard"],"Server":["TheRock/2.0"],"Vary":["Accept"],"Content-Type":["text/plain; charset=utf-8"],"Content-Length":["32"]}}
```

where the `%K` should be `TLSv1.3` and `%k` should be `TLS_AES_128_GCM_SHA256`.

But the ENCRYPTION SETTINGS panel in the real-time HTML is still not showing any data while all the other panels show data.

Any idea how to fix this? Has the Caddy JSON format maybe changed since the last update of the `--log-format=CADDY` definition?

**Comment by allinurl** [comment]
Thank you for sharing this. Do you happen to know if this is the default field output or if it has been changed at some point? The current output you provided is:

    "version":"TLSv1.3","cipher_suite": "TLS_AES_128_GCM_SHA256"

However, the default in GoAccess includes "cipher_suite" and "proto". In any event, the following should still function properly:

    goaccess access.log --log-format='{ "ts": "%x.%^", "request": { "remote_ip": "%h", "proto": "%H", "method": "%m", "host": "%v", "uri": "%U", "headers": { "User-Agent": [ "%u" ], "Referer": [ "%R" ] }, "tls": { "cipher_suite": "%k", "version": "%K" } }, "duration": "%T", "size": "%b", "status": "%s", "resp_headers": { "Content-Type": [ "%M" ] } }' --date-format=%s --time-format=%s

**Comment by steffenbusch** [comment]
> However, the default in GoAccess includes "cipher_suite" and "proto". In any event, the following should still function properly:
> 
> ```
> goaccess access.log --log-format='{ "ts": "%x.%^", "request": { "remote_ip": "%h", "proto": "%H", "method": "%m", "host": 
> ```

Thank your for your help @allinurl. 
**That solved the problem.**

The first JSON posted above is the default field output of Caddy v2.7.4.  with
```json
"tls":{"resumed":false,"version":772,"cipher_suite":4865,"proto":"http/1.1","server_name":"arm.stbu.net"}
```

Using the GoAccess v1.8.0 with `--caddy-format=CADDY" wasn't working. 
You might want to reproduce that using:

```bash
$ /home/stbu/goaccess --version | grep GoAccess
GoAccess - 1.8.
$ echo '{"level":"info","ts":1696325286.9971106,"logger":"http.log.access.arm.stbu.net","msg":"handled request","request":{"remote_ip":"46.101.122.201","remote_port":"39062","client_ip":"46.101.122.201","proto":"HTTP/1.1","method":"GET","host":"arm.stbu.net","uri":"/","headers":{"Accept-Encoding":["gzip, deflate"],"User-Agent":["OhDear.app (+https://ohdear.app/docs/checks/uptime)"],"Accept":["*/*"]},"tls":{"resumed":false,"version":772,"cipher_suite":4865,"proto":"http/1.1","server_name":"arm.stbu.net"}},"bytes_read":0,"user_id":"","duration":0.004788534,"size":32,"status":302,"resp_headers":{"Content-Type":["text/plain; charset=utf-8"],"Content-Length":["32"],"Date":["Tue, 03 Oct 2023 09:28:06 GMT"],"X-Frame-Options":["SAMEORIGIN"],"Location":["/dashboard"],"Server":["TheRock/2.0"],"Vary":["Accept"]}}' | \
/home/stbu/goaccess --log-format=CADDY  -o /tmp/goaccess-caddy.html -
```

Since the values of `772` for the TLS version and `4865` for the cipher-suite are not very user-friendly in the default Caddy JSON as well as in the GoAccess HTML report (if they would show up), I have modified the Caddy JSON log configuration like this:

```txt
       log access-json {
                include http.log.access
                output file log/access.json
                format filter {
                        wrap json
                        fields {
                                request>tls>version tls_version TLSv
                                request>tls>cipher_suite tls_cipher
                        }
                }
        }
```

and that produces this human friendly output (but the field labels are the same, just now strings instead of integers)

```json
"tls":{"resumed":false,"version":"TLSv1.3","cipher_suite":"TLS_AES_128_GCM_SHA256","proto":"http/1.1","server_name":"arm.stbu.net"}
```

If I use your suggested `--log-format` it's working for me:

```bash
echo '{"level":"info","ts":1696324268.7687688,"logger":"http.log.access.arm.stbu.net","msg":"handled request","request":{"remote_ip":"167.71.54.244","remote_port":"36770","client_ip":"167.71.54.244","proto":"HTTP/1.1","method":"GET","host":"arm.stbu.net","uri":"/","headers":{"Accept-Encoding":["gzip, deflate"],"User-Agent":["OhDear.app (+https://ohdear.app/docs/checks/uptime)"],"Accept":["*/*"]},"tls":{"resumed":false,"version":"TLSv1.3","cipher_suite":"TLS_AES_128_GCM_SHA256","proto":"http/1.1","server_name":"arm.stbu.net"}},"bytes_read":0,"user_id":"","duration":0.003010369,"size":32,"status":302,"resp_headers":{"Content-Length":["32"],"Date":["Tue, 03 Oct 2023 09:11:08 GMT"],"Server":["TheRock/2.0"],"X-Frame-Options":["SAMEORIGIN"],"Location":["/dashboard"],"Vary":["Accept"],"Content-Type":["text/plain; charset=utf-8"]}}' | \
/home/stbu/goaccess --log-format='{ "ts": "%x.%^", "request": { "remote_ip": "%h", "proto": "%H", "method": "%m", "host": "%v", "uri": "%U", "headers": { "User-Agent": [ "%u" ], "Referer": [ "%R" ] }, "tls": { "cipher_suite": "%k", "version": "%K" } }, "duration": "%T", "size": "%b", "status": "%s", "resp_headers": { "Content-Type": [ "%M" ] } }' --date-format=%s --time-format=%s -o /tmp/goaccess-caddy.html -
```

So basically, the `--log-format=CADDY` is either not capturing the `%K` and `%k` properly (maybe because the JSON fields have changed since its inception), or the integer values of e.g. `772` for the TLS version and e.g. `4865` for cipher-suite are not recognized and shown by GoAccess.
However, I am very happy with the modified Caddy JSON and your suggested solution of `--log-format='{ "ts": "%x.%^", "request": { "remote_ip": "%h", "proto": "%H", "method": "%m", "host": "%v", "uri": "%U", "headers": { "User-Agent": [ "%u" ], "Referer": [ "%R" ] }, "tls": { "cipher_suite": "%k", "version": "%K" } }, "duration": "%T", "size": "%b", "status": "%s", "resp_headers": { "Content-Type": [ "%M" ] } }' --date-format=%s --time-format=%s`

**Comment by PotatoMaaan** [comment]
Hi, I'm having the same issue. My caddy logging is configured like this:
```caddy
log {
     output file /var/log/caddy/access_combined.log {
         roll_size 2GiB
    }
    format json
    level INFO
}
```
I'm using the default json output, so I would expect `--log-format=caddy` to pick up on the tls information provided by the default caddy output.

As already stated above, the output tls information looks like this (formatted to look nice):
```json
"tls": {
    "resumed": false,
    "version": 772,
    "cipher_suite": 4865,
    "proto": "h2",
    "server_name": "[REDACTED]"
}
```

Using the custom log format string with goaccess also works for me, but I think it would be best if the default caddy format string in goaccess would be updated to reflect the caddy default output.

**Comment by timelordx** [comment]
Experiencing the exact same issue. 

Caddy v2.10.0
GoAccess v1.9.4

```yaml
services:
  goaccess:
    image: allinurl/goaccess:latest
    container_name: goaccess
    environment:
      - TZ=PST8PDT
    volumes:
      - $PWD/goaccess/db:/db
      - $PWD/caddy/logs:/logs:ro
      - $PWD/caddy/www/stats:/www
      - $PWD/maxmind:/maxmind:ro
    command:
      - '--geoip-database=/maxmind/GeoLite2-City.mmdb'
      - '--log-file=/logs/access.log'
      - '--log-format=CADDY'
      - '--db-path=/db'
      - '--persist'
      - '--restore'
      - '--real-time-html'
      - '--output=/www/index.html'
      - '--ws-url=wss://webstats.example.com:443'
    ports:
      - 127.0.0.1:7890:7890
    restart: unless-stopped
```
My `Caddyfile` log section looks like this:
```
log {
	output file /logs/access.log {
		roll_size 100mb
		roll_keep 5
	}
	format json
}
```
Everything seems to be running fine, but the `Encryption Settings` panel shows up empty:

<img width="1163" alt="Image" src="https://github.com/user-attachments/assets/41c98874-06b1-482c-947a-d59270e1c493" />

Here’s a sample from the TLS section of the log output (formatted for readability):

```json
"tls": {
  "resumed": true,
  "version": 772,
  "cipher_suite": 4865,
  "proto": "h3",
  "server_name": "webstats.example.com"
}
```

**Comment by allinurl** [comment]
@timelordx, did you use `--with-openssl` when building? If you could post the complete JSON line, I'd appreciate it. Want to see if I can reproduce the issue on a vanilla setup.

**Comment by timelordx** [comment]
Thanks for getting back to me, @allinurl.

I'm using your `allinurl/goaccess:latest` Docker image, which appears to be compiled with OpenSSL:
```bash
$ sudo docker exec -ti goaccess goaccess -V
GoAccess - 1.9.4.
For more details visit: https://goaccess.io/
Copyright (C) 2009-2024 by Gerardo Orellana

Build configure arguments:
  --enable-geoip=mmdb
  --with-openssl
```
You can see a full setup example of my deployment here:

https://caddy.community/t/goaccess-log-file-analysis-for-caddy/31246

Here’s the output from `tail -n1 access.log`:
```json
{"level":"info","ts":1750181395.9012785,"logger":"http.log.access.log1","msg":"handled request","request":{"remote_ip":"192.168.1.20","remote_port":"57404","client_ip":"192.168.1.20","proto":"HTTP/3.0","method":"GET","host":"www.example.com","uri":"/api/v1/status","headers":{"Sec-Fetch-Mode":["cors"],"Accept-Language":["en-CA,en-US;q=0.9,en;q=0.8"],"Accept-Encoding":["gzip, deflate, br"],"Cookie":["REDACTED"],"Referer":["https://www.example.com/"],"Priority":["u=3, i"],"Sec-Fetch-Dest":["empty"],"User-Agent":["Mozilla/5.0 (iPhone; CPU iPhone OS 18_5 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.5 Mobile/15E148 Safari/604.1"],"Accept":["application/json, text/plain, */*"],"Sec-Fetch-Site":["same-origin"]},"tls":{"resumed":false,"version":772,"cipher_suite":4865,"proto":"h3","server_name":"www.example.com"}},"bytes_read":0,"user_id":"","duration":1.475911583,"size":108,"status":200,"resp_headers":{"Alt-Svc":["h3=\":443\"; ma=2592000"],"Strict-Transport-Security":["max-age=16070400"],"Content-Length":["108"],"Content-Security-Policy":["upgrade-insecure-requests"],"X-Content-Type-Options":["nosniff"],"Content-Type":["application/json; charset=utf-8"],"Etag":["W/\"6c-OwgZPtwZz6mQh9HUvHGrbyuWVUk\""],"Date":["Tue, 17 Jun 2025 17:29:55 GMT"],"X-Xss-Protection":["1; mode=block"],"Referrer-Policy":["no-referrer, strict-origin-when-cross-origin"],"X-Frame-Options":["DENY"]}}
```
I've replaced the hostname with `www.example.com`. 

I’m running a nearly vanilla Caddy setup, using ACMEDNS module for DNS-01 certificate issuance:

https://github.com/timelordx/caddy-dns-acmedns/blob/main/Dockerfile

Happy to provide more details if needed!

**Comment by allinurl** [comment]
Sorry for the delay. I do get with v1.9.4 for that sample line you posted:

![Image](https://github.com/user-attachments/assets/8f8eba66-c7c5-47e8-8fa8-a9ae738ea33f)

I'm curious, what version of libssl is currently running?

**Comment by timelordx** [comment]
No worries at all. I totally understand your schedule can get pretty busy.

> I'm curious, what version of libssl is currently running?

If you let me know which app you're referring to and how I can find that info, I’ll happily check it for you.

For context, I’m using this image:

```
$ sudo docker inspect --format='{{index .RepoDigests 0}}' allinurl/goaccess
allinurl/goaccess@sha256:834b53b7e6df0836d9c8742c7cc7276e1b2405ce4fc56d46f14e6de12668945b
```
It’s currently the [allinurl/goaccess:latest](https://hub.docker.com/layers/allinurl/goaccess/latest/images/sha256-bbfbf8b9ee833ef3a9f34a812a7ce9557e4980cbd146c781259c3e0d23813448) `linux/amd64` image on Docker Hub.

**Comment by timelordx** [comment]
If it helps, here are a couple of tests I ran:

**access.log:**
```
{"level":"info","ts":1750181395.9012785,"logger":"http.log.access.log1","msg":"handled request","request":{"remote_ip":"192.168.1.20","remote_port":"57404","client_ip":"192.168.1.20","proto":"HTTP/3.0","method":"GET","host":"www.example.com","uri":"/api/v1/status","headers":{"Sec-Fetch-Mode":["cors"],"Accept-Language":["en-CA,en-US;q=0.9,en;q=0.8"],"Accept-Encoding":["gzip, deflate, br"],"Cookie":["REDACTED"],"Referer":["https://www.example.com/"],"Priority":["u=3, i"],"Sec-Fetch-Dest":["empty"],"User-Agent":["Mozilla/5.0 (iPhone; CPU iPhone OS 18_5 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.5 Mobile/15E148 Safari/604.1"],"Accept":["application/json, text/plain, */*"],"Sec-Fetch-Site":["same-origin"]},"tls":{"resumed":false,"version":772,"cipher_suite":4865,"proto":"h3","server_name":"www.example.com"}},"bytes_read":0,"user_id":"","duration":1.475911583,"size":108,"status":200,"resp_headers":{"Alt-Svc":["h3=\":443\"; ma=2592000"],"Strict-Transport-Security":["max-age=16070400"],"Content-Length":["108"],"Content-Security-Policy":["upgrade-insecure-requests"],"X-Content-Type-Options":["nosniff"],"Content-Type":["application/json; charset=utf-8"],"Etag":["W/\"6c-OwgZPtwZz6mQh9HUvHGrbyuWVUk\""],"Date":["Tue, 17 Jun 2025 17:29:55 GMT"],"X-Xss-Protection":["1; mode=block"],"Referrer-Policy":["no-referrer, strict-origin-when-cross-origin"],"X-Frame-Options":["DENY"]}}
```

**macOS 15.5 Homebrew**
```
$ brew install goaccess
```
```
$ goaccess -V
GoAccess - 1.9.4.
For more details visit: https://goaccess.io/
Copyright (C) 2009-2024 by Gerardo Orellana

Build configure arguments:
  --enable-geoip=mmdb
```
```
$ goaccess access.log --output=homebrew.html --log-format=CADDY
```
**Docker Image allinurl/goaccess:latest**
```
$ docker run -ti --rm -v $PWD:/goaccess allinurl/goaccess:latest -V
GoAccess - 1.9.4.
For more details visit: https://goaccess.io/
Copyright (C) 2009-2024 by Gerardo Orellana

Build configure arguments:
  --enable-geoip=mmdb
  --with-openssl
```
```
$ docker run -ti --rm -v $PWD:/goaccess allinurl/goaccess:latest /goaccess/access.log --output=/goaccess/latest.html --log-format=CADDY
```
**Docker Image allinurl/goaccess:1.9.4**
```
$ docker run -ti --rm -v $PWD:/goaccess allinurl/goaccess:1.9.4 -V
GoAccess - 1.9.4.
For more details visit: https://goaccess.io/
Copyright (C) 2009-2024 by Gerardo Orellana

Build configure arguments:
  --enable-geoip=mmdb
  --with-openssl
```
```
$ docker run -ti --rm -v $PWD:/goaccess allinurl/goaccess:1.9.4 /goaccess/access.log --output=/goaccess/194.html --log-format=CADDY
```
All of these tests gave me this result:

<img width="763" alt="Image" src="https://github.com/user-attachments/assets/4e19e740-8311-4427-88b5-569e9e02e98e" />

Everything else seems to work fine. If I’ve overlooked something or made a mistake, I’d really appreciate any guidance!

**Comment by allinurl** [comment]
Thanks for sharing that. Looks like a docker-related issue, as I'm seeing the same issue. I'll dig a bit deeper and follow up once I find something useful.

**Comment by allinurl** [comment]
@timelordx I've updated the Dockerfile, it should be working now. The issue was that LibreSSL doesn't support the SSL_CIPHER_standard_name() function, which is part of the OpenSSL 3.x API used for resolving the cipher. You may need to clear the cache and pull it again. Let me know how that goes.

**Comment by allinurl** [comment]
@timelordx rushed on my response, let me fix a build issue and will post back.

**Comment by allinurl** [comment]
Give it a shot, should work now. Let me know if that works for you.

**Comment by timelordx** [comment]
While the Homebrew version is still giving me an empty "Encryption Settings" panel, the `allinurl/goaccess:latest` Docker image is now showing me the data! I'm calling it a success! 🎉

<img width="870" alt="Image" src="https://github.com/user-attachments/assets/3229481c-8c84-41b1-a88a-78503dc45e30" />


Thank you for jumping on it and taking care of it so quickly! 🙇🏻‍♂️

**Comment by allinurl** [comment]
Glad that solved the issue. The homebrew seems to be missing building using —-with-openssl. They will need to update the formula.  

https://github.com/Homebrew/homebrew-core/blob/87c475073d7bf4ac84fd39e7dee17813dee830e7/Formula/g/goaccess.rb#L32

**Comment by allinurl** [comment]
@timelordx thank you by the way for reporting this, much appreciated.
